### PR TITLE
[SPARK-40370][SQL] Migrate type check fails to error classes in CAST

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to use error classes in the case of type check failure in the `CAST` expression.

### Why are the changes needed?
Migration onto error classes unifies Spark SQL error messages, and improves search-ability of errors.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
By running the modified test suites:
```
$ build/sbt "test:testOnly *CastWithAnsiOnSuite"
$ build/sbt "test:testOnly *DatasetSuite"
$ build/sbt "sql/testOnly org.apache.spark.sql.SQLQueryTestSuite -- -z cast.sql"
```

### CODE REVIEW

**Comment by MaxGekk** [comment]
@srielau @anchovYu @cloud-fan @gengliangwang Could you review this PR, please.

File: sql/core/src/test/resources/sql-tests/results/ansi/cast.sql.out [hunk 1]
```diff
@@ -611,10 +611,26 @@ SELECT HEX(CAST(CAST(123 AS byte) AS binary))
...
-; line 1 pos 11
+{
+  "errorClass" : "DATATYPE_MISMATCH",
+  "errorSubClass" : "CAST_WITH_CONF_SUGGESTION",
+  "messageParameters" : {
+    "config" : "\"spark.sql.ansi.enabled\"",
+    "configVal" : "'false'",
+    "sqlExpr" : "\"CAST(CAST(123 AS TINYINT) AS BINARY)\"",
+    "srcType" : "\"TINYINT\"",
+    "srcType" : "\"TINYINT\"",
```
Comment (MaxGekk) [inline sql/core/src/test/resources/sql-tests/results/ansi/cast.sql.out:622]: We have a bug when an error message refers to the same param more than once. I will fix this separately.

**Review by cloud-fan — APPROVED** [review]

**Comment by MaxGekk** [comment]
The failure "SQLAppStatusListenerMemoryLeakSuite.no memory leak" is not related to the changes, I believe.
Merging to master. Thank you, @cloud-fan for review.

File: core/src/main/resources/error/error-classes.json [hunk 2]
```diff
@@ -95,6 +95,23 @@
...
+          "cannot cast <srcType> to <targetType>."
+        ]
+      },
+      "CAST_WITH_CONF_SUGGESTION" : {
+        "message" : [
+          "cannot cast <srcType> to <targetType> with ANSI mode on.",
+          "If you have to cast <srcType> to <targetType>, you can set <config> as <configVal>."
+        ]
+      },
+      "CAST_WITH_FUN_SUGGESTION" : {
```
Comment (cloud-fan) [inline core/src/main/resources/error/error-classes.json:110]: nit: `FUN` is really confusing here. I think `FUNC` is more commonly used as a short name of FUNCTION

Comment (MaxGekk) [reply core/src/main/resources/error/error-classes.json:110]: Tell this Rust programmers ;-)

**Pull request closed without merging**